### PR TITLE
Initial codeblock syntax highlighting support

### LINKED ISSUE #8: Support codeblock highlighting
It would be really nice to have syntax highlighting support for `inlyne`. I haven't delved into the html rendering side, but getting `comrak` to support adding colored spans in codeblocks is as simple as

```patch
diff --git a/src/main.rs b/src/main.rs
index c14b060..7677e81 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -928,7 +928,12 @@ fn main() {
         options.extension.strikethrough = true;
         options.parse.smart = true;
         options.render.unsafe_ = true;
-        let htmlified = markdown_to_html(&md_string, &options);
+
+        let mut plugins = comrak::ComrakPlugins::default();
+        let adapter = comrak::plugins::syntect::SyntectAdapter::new("base16-ocean.dark");
+        plugins.render.codefence_syntax_highlighter = Some(&adapter);
+
+        let htmlified = comrak::markdown_to_html_with_plugins(&md_string, &options, &plugins);
 
         input.push_back(
             Tendril::from_str(&htmlified)
```

The theme is configurable (passed to `SyntextAdapter`), and this renders html that looks like so when viewed in my browser

image

and here is the raw HTML for the first block

```html
<pre style="background-color:#2b303b;"><code class="language-js">
<span style="color:#65737e;">// cat.js
</span><span style="color:#b48ead;">import </span><span style="color:#c0c5ce;">{ </span><span style="color:#bf616a;">resolve </span><span style="color:#c0c5ce;">} </span><span style="color:#b48ead;">from </span><span style="color:#c0c5ce;">&quot;</span><span style="color:#a3be8c;">path</span><span style="color:#c0c5ce;">&quot;;
</span><span style="color:#b48ead;">import </span><span style="color:#c0c5ce;">{ </span><span style="color:#bf616a;">write</span><span style="color:#c0c5ce;">, </span><span style="color:#bf616a;">stdout</span><span style="color:#c0c5ce;">, </span><span style="color:#bf616a;">file</span><span style="color:#c0c5ce;">, </span><span style="color:#bf616a;">argv </span><span style="color:#c0c5ce;">} </span><span style="color:#b48ead;">from </span><span style="color:#c0c5ce;">&quot;</span><span style="color:#a3be8c;">bun</span><span style="color:#c0c5ce;">&quot;;
</span><span style="color:#c0c5ce;">
</span><span style="color:#b48ead;">const </span><span style="color:#c0c5ce;">path = </span><span style="color:#8fa1b3;">resolve</span><span style="color:#c0c5ce;">(</span><span style="color:#bf616a;">argv</span><span style="color:#c0c5ce;">.</span><span style="color:#8fa1b3;">at</span><span style="color:#c0c5ce;">(-</span><span style="color:#d08770;">1</span><span style="color:#c0c5ce;">));
</span><span style="color:#c0c5ce;">
</span><span style="color:#b48ead;">await </span><span style="color:#8fa1b3;">write</span><span style="color:#c0c5ce;">(
</span><span style="color:#c0c5ce;">  </span><span style="color:#65737e;">// stdout is a Blob
</span><span style="color:#c0c5ce;">  </span><span style="color:#bf616a;">stdout</span><span style="color:#c0c5ce;">,
</span><span style="color:#c0c5ce;">  </span><span style="color:#65737e;">// file(path) returns a Blob - https://developer.mozilla.org/en-US/docs/Web/API/Blob
</span><span style="color:#c0c5ce;">  </span><span style="color:#8fa1b3;">file</span><span style="color:#c0c5ce;">(path)
</span><span style="color:#c0c5ce;">);
</span><span style="color:#c0c5ce;">
</span><span style="color:#65737e;">// bun ./cat.js ./path-to-file
</span>
</code></pre>
```

I would be happy to do the work rendering the colored spans if you think this change would be worthwhile

### PR DESCRIPTION
**Pull request by LovecraftianHorror**
Resolves #8

_The `comrak` feature tweaking is to get rid of the default `clap` feature which is only useful for the `comrak` binary_

This implements the initial support for syntax highlighting code blocks, but it seems there are some bugs in how some spans with just whitespace are handled (e.g. the lines starting with `async` and `return`) which I would appreciate some help troubleshooting since the `TokenPrinter` is a big ball of mutable state that is pretty hard to follow.

Here is some of the syntax highlighted code where the indentation is having some obvious issues. The indentation issues are reproducible when just having `comrak` include colored spans in codeblocks without any of the code for respecting the span's color

_Before_

image

_After_

image

---

Also worth noting that the HTML portion of the code snippet should also be indented, but isn't in either this PR or `main`

```js
// requires Bun v0.1.0 or later
// react-ssr.tsx
import { renderToReadableStream } from "react-dom/server";

const dt = new Intl.DateTimeFormat();

export default {
  port: 3000,
  async fetch(request: Request) {
    return new Response(
      await renderToReadableStream(
        <html>
          <head>
            <title>Hello World</title>
          </head>
          <body>
            <h1>Hello from React!</h1>
            <p>The date is {dt.format(new Date())}</p>
          </body>
        </html>
      )
    );
  },
};

// bun react-ssr.tsx
```

### CODE REVIEW

**Comment by trimental** [comment]
I tried in c746cbdc3ab9d7d82c51b1dcf8d335cfac41717a to clean up the html interpreter a bit as it is definitely a mess 😅. To briefly explain it now uses a `State` struct to track things like span color, text formatting options, links, etc. For some things I've stored this as an integer instead of boolean to track levels of depth, ie

```
<a href="cats.html">
<a href="dogs.html">
<p> a link </p> // Depth 2
</a>
<p> still a link </p> // Depth 1
</a>
// Depth 0
```

While element_stack stores elements that their children need access to, so <table> is stored on this stack so <td> <th> <tr> can access it.

About your the whitespace not being accounted for, its because wgpu_glyph only reports whitespaces between text and not at the start or ends so that the size is representative of text and not whitespace.

I've managed to get this
Screenshot 2022-08-15 at 12 46 18 am
when adding 
```
                if str == "\n" {
                    if self.state.text_options.pre_formatted >= 1 {
                        if !self.current_textbox.texts.is_empty() {
                            self.push_element(self.current_textbox.clone().into());
                            self.current_textbox.texts.clear();
                        }
                    } else if !self.current_textbox.texts.is_empty() {
                        self.current_textbox.texts.push(Text::new(
                            " ".to_string(),
                            self.hidpi_scale,
                            self.theme.text_color,
                        ));
                    }
                } else {
```
 to https://github.com/trimental/inlyne/commit/c746cbdc3ab9d7d82c51b1dcf8d335cfac41717a#diff-60405aaba3dc351a041fce3e90f17b63dee9a214bc9c35ae6a5126d003100e55R448

**Comment by trimental** [comment]
By the way code colouring makes the application feel a lot more complete, once again thank you for taking the time to help out with this.

File: src/main.rs [hunk 1]
```diff
@@ -686,6 +688,27 @@ impl TokenSink for TokenPrinter {
...
+                                    if let Some(hex_str) = styles
+                                        .split(';')
+                                        .find_map(|style| style.strip_prefix("color:#"))
+                                    {
+                                        if let Ok(hex) = u32::from_str_radix(hex_str, 16) {
+                                            let [_, r, g, b] = hex.to_be_bytes();
+                                            let r = r as f32 / 255.0;
+                                            let g = g as f32 / 255.0;
+                                            let b = b as f32 / 255.0;
+                                            self.span_color = [r, g, b, 1.0];
```
Comment (trimental) [inline src/main.rs:706]: Using the function `hex_to_linear_rgba` from colour.rs like `self.state.span_color = hex_to_linear_rgba(hex);` helps to convert the srgb web colour into rgba linear which wgpu uses. We need to do this conversion to stop colors washing out and appearing muted.

File: src/main.rs [hunk 2]
```diff
@@ -916,7 +938,12 @@ fn main() -> anyhow::Result<()> {
         options.extension.strikethrough = true;
         options.parse.smart = true;
         options.render.unsafe_ = true;
-        let htmlified = markdown_to_html(&md_string, &options);
+
+        let mut plugins = comrak::ComrakPlugins::default();
+        let adapter = comrak::plugins::syntect::SyntectAdapter::new("InspiredGitHub");
```
Comment (trimental) [inline src/main.rs:943]: It would also be nice to be able to change this depending on the theme. It should be as simple as adding `pub code_highlighter: &'static str` to `Theme` and replacing `InspiredGithub` with `theme.code_highlighter`.

I think `InspiredGithub` looks really good for light theme, and I'm a fan of `base16-mocha.dark`. Please let me know if you have any other suggestions.

**Comment by LovecraftianHorror** [comment]
Thanks for the explanation on how the html interpreter works, it really helped clear things up a lot!

If you already have a branch that has the codeblocks highlighting changes on top of the html interpreter refactor then feel free to push it here and I'll address the other review comments!

File: src/main.rs (hunk 2 above)
Comment (LovecraftianHorror) [reply src/main.rs:943]: That sounds good to me!

Eventually we can make it configurable so that users can pick any of the themes that `syntect` supports. `comrak` doesn't expose a way to set a theme other than the ones included by default AFAICT, but that still give users a handful of different options at least

**Comment by trimental** [comment]
@LovecraftianHorror I've just forced my branch of your PR for you to take a look at on your device. 

I also added a padding function to `TextBox` that just adds height regardless of whether the text is whitespace. By adding a textbox with a padding of 18., it's the same as adding an empty line. This fixes the fact that the code boxes used to ignore empty lines.

**Comment by LovecraftianHorror** [comment]
Just pulled it down, and it looks great!

Thanks for helping out so much with this PR. It looks like that resolves both of the review comments

**Comment by trimental** [comment]
@LovecraftianHorror Team work makes the dream work. I'll spend tomorrow looking for bugs and hopefully release the next version tomorrow night 👍

**Comment by LovecraftianHorror** [comment]
Sounds great!

Most of my contributing will be bursts on weekends, so I won't try pushing through more changes before the next release :smile:
